### PR TITLE
[Debt] Adds Playwright clock

### LINKED ISSUE #10819: ♻️ Implement Playwright clock in login-logout tests
## ♻️ Debt/Refactor

When [the login-logout tests](https://github.com/GCTC-NTGC/gc-digital-talent/blob/main/apps/playwright/tests/login-logout.spec.ts) were migrated from Cyrpess to Playwright, Playwright had no clock API of its own and so a custom helper was developed.  Now that Playwright has this feature builtin the custom implementation should be removed.

## 🕵️ Details

https://playwright.dev/docs/api/class-clock

## ✅ Acceptance Criteria

- [ ] https://github.com/GCTC-NTGC/gc-digital-talent/blob/main/apps/playwright/utils/clock.ts removed
- [ ] Sinon JS dependency removed
- [ ] login-logout test updated to use Playwright clock

### PR DESCRIPTION
**Pull request by mnigh**
🤖 Resolves #10819.

## 👋 Introduction

This PR adds the Playwright clock API and removes the custom helper that was developed before the Playwright clock API existed.

## :detective: Details

It seems the playwright clock works slightly differently from `sinon` so some changes had to occur. There was a refactor to use a fixture to avoid needing to update refresh step individually. 

This test was a little hard to follow so I tried to document the refresh step in order to explain what is happening and why. Hope it helps.

## 🧪 Testing

Assist reviewers with steps they can take to test that the PR does what it says it does.

1. `pnpm e2e:playwright:chromium -- login-logout.spec.ts`
2. Ensure replacement clock functions make sense

### CODE REVIEW

**Review by petertgiles — COMMENTED** [review]
Very cool!  I just did a readthrough and am trying to understand it.  I left a few questions and suggestions to consider.

File: apps/playwright/fixtures/AuthTokenFixture.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    const authReqPromise = this.page.waitForRequest(async (req) => {
+      if (req.url()?.includes("/graphql")) {
+        const reqJson = (await req?.postDataJSON()) as GraphQLOperation | null;
+        return typeof reqJson?.operationName !== "undefined";
+      }
+      return false;
+    });
+
+    // Reload page to force the token refresh
+    await this.page.reload();
```
[the remark this answers is not in the document]
Comment (esizer) [reply apps/playwright/fixtures/AuthTokenFixture.ts:52]: In order to intercept the refresh, it needs to happen in context of the browser where the request happens. We want the currently logged in user to cause the refresh and get the new tokens from the browser. 

The only way to do this consistency is to use the browser itself and intercept the refresh. Also, the reload ensures a graphql request is made. `urql` caches results so navigating to the page does not guarantee a request is made. Reload, while slower, ensures the `urql` cache is purged and we are guaranteed to get a request to fire.

File: apps/playwright/fixtures/AuthTokenFixture.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    // Get the token used to refresh from search params
+    const refreshTokenUsed = await refreshPromise.then((req) =>
+      new URL(req.url()).searchParams.get("refresh_token"),
+    );
+
+    // Reset clock to avoid causing refresh issues
+    // NOTE: Not sure but think the auth server clock is out of sync?
+    await this.page.clock.setSystemTime(new Date());
+    // Reload page to force a graphql request with new tokens
+    await this.page.reload();
```
[the remark this answers is not in the document]
Comment (esizer) [reply apps/playwright/fixtures/AuthTokenFixture.ts]: Huh, idk why I thought that was needed :sweat_smile: Removed in [remove unecessary reload](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10830/commits/8dadb1e2ecec0beb2f986d7e4aaec81939c26c75)

File: apps/playwright/fixtures/AuthTokenFixture.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  async jumpPastExpiry(tokens?: AuthTokens) {
+    await this.page.clock.setSystemTime(
+      jumpPastExpiryDate(tokens?.accessToken ?? ""),
+    );
+  }
+
+  /**
+   * Force a token refresh and store the new tokens for assertions
+   */
+  async forceRefreshAndGetNewTokens(): Promise<RefreshedTokens> {
```
[the remark this answers is not in the document]
Comment (esizer) [reply apps/playwright/fixtures/AuthTokenFixture.ts]: Okay, that makes sense. I hide the implementation details in the fixture but broke it up significantly so you can see the steps being taken in the test itself.

[refactor fixture to not be so magic](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10830/commits/5a85a85fdf984f3a59a6102f5b6f7deeb6ec73c2)

File: apps/playwright/fixtures/AuthTokenFixture.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    await this.page.reload();
+
+    // Get the token used to refresh from search params
+    const refreshTokenUsed = await refreshPromise.then((req) =>
+      new URL(req.url()).searchParams.get("refresh_token"),
+    );
+
+    // Reset clock to avoid causing refresh issues
+    // NOTE: Not sure but think the auth server clock is out of sync?
+    await this.page.clock.setSystemTime(new Date());
```
[the remark this answers is not in the document]
Comment (esizer) [reply apps/playwright/fixtures/AuthTokenFixture.ts]: Unfortunately there is not reset :cry: Though, I think the date is maybe not correct. In the refactor I updated this to use `Date.now()` instead which I pulled from [the documentation](https://playwright.dev/docs/api/class-clock#clock-set-system-time:~:text=await%20page.clock.setSystemTime(Date.now())%3B).

File: apps/playwright/tests/login-logout.spec.ts [hunk 5]
```diff
@@ -171,101 +163,83 @@ test.describe("Login and logout", () => {
...
-        );
-      });
+    // Third refresh for good measure
+    await fixture.jumpPastExpiry(tokenSet3);
 
-    // reset clock
-    await clockHelper.restore();
-    // time travel to when the tokens from token set 3 expire before trying to navigate
-    await clockHelper.jumpTo(jumpPastExpiryDate(tokenSet3?.accessToken ?? ""));
+    const listenters3 = fixture.createListeners();
```
Comment (mnigh) [inline apps/playwright/tests/login-logout.spec.ts:231]:
```suggestion
    const listeners3 = fixture.createListeners();
```

File: apps/playwright/tests/login-logout.spec.ts [hunk 6]
```diff
@@ -171,101 +163,83 @@ test.describe("Login and logout", () => {
...
-    const request3 = await requestPromise;
-    // navigate to a page
-    await page.goto("/en/applicant");
-    // get refresh token 3 from request URL
-    const refreshToken3 = new URL(request3.url()).searchParams.get(
-      "refresh_token",
+    await fixture.page.reload();
+
+    const refreshToken3 = await fixture.getRefreshTokenUsed(
+      listenters3.refresh,
```
Comment (mnigh) [inline apps/playwright/tests/login-logout.spec.ts:236]:
```suggestion
      listeners3.refresh,
```

File: apps/playwright/tests/login-logout.spec.ts [hunk 7]
```diff
@@ -171,101 +163,83 @@ test.describe("Login and logout", () => {
...
-    // navigate to a page
-    await page.goto("/en/applicant");
-    // get refresh token 1 from request 1 URL
-    const refreshToken1 = new URL(request.url()).searchParams.get(
-      "refresh_token",
+    // Time travel past the expiry darte of token set 1
+    await fixture.jumpPastExpiry(tokenSet1);
+
+    // Setup request listeners
+    const listenters1 = fixture.createListeners();
```
Comment (mnigh) [inline apps/playwright/tests/login-logout.spec.ts:181]:
```suggestion
    const listeners1 = fixture.createListeners();
```

File: apps/playwright/tests/login-logout.spec.ts [hunk 8]
```diff
@@ -171,101 +163,83 @@ test.describe("Login and logout", () => {
...
+
+    // Reload page to force refresh
+    await fixture.page.reload();
+
+    // Get refresh token used
+    const refreshToken1 = await fixture.getRefreshTokenUsed(
+      listenters1.refresh,
```
Comment (mnigh) [inline apps/playwright/tests/login-logout.spec.ts:188]:
```suggestion
      listeners1.refresh,
```

File: apps/playwright/tests/login-logout.spec.ts [hunk 9]
```diff
@@ -171,101 +163,83 @@ test.describe("Login and logout", () => {
...
     );
+
+    // Reset clock to avoid unecessary refreshes
+    await fixture.resetClock();
+
+    const authHeader1 = await fixture.getAuthHeader(listenters1.graphql);
```
Comment (mnigh) [inline apps/playwright/tests/login-logout.spec.ts:194]:
```suggestion
    const authHeader1 = await fixture.getAuthHeader(listeners1.graphql);
```

File: apps/playwright/tests/login-logout.spec.ts [hunk 10]
```diff
@@ -171,101 +163,83 @@ test.describe("Login and logout", () => {
...
-      })
-      .then((req) => {
-        // make sure it uses the second access token
-        expect(req.headers().authorization).toEqual(
-          `Bearer ${tokenSet2.accessToken}`,
-        );
-      });
+    //  Repeat steps with the second token set
+    await fixture.jumpPastExpiry(tokenSet2);
+    const listenters2 = fixture.createListeners();
```
Comment (mnigh) [inline apps/playwright/tests/login-logout.spec.ts:207]:
```suggestion
    const listeners2 = fixture.createListeners();
```

File: apps/playwright/tests/login-logout.spec.ts [hunk 11]
```diff
@@ -171,101 +163,83 @@ test.describe("Login and logout", () => {
...
+    await fixture.page.reload();
 
-    const request2 = await requestPromise;
-    // navigate to a page
-    await page.goto("/en/applicant");
-    // get refresh token 2 from request URL
-    const refreshToken2 = new URL(request2.url()).searchParams.get(
-      "refresh_token",
+    const refreshToken2 = await fixture.getRefreshTokenUsed(
+      listenters2.refresh,
```
Comment (mnigh) [inline apps/playwright/tests/login-logout.spec.ts:212]:
```suggestion
      listeners2.refresh,
```

File: apps/playwright/tests/login-logout.spec.ts [hunk 12]
```diff
@@ -171,101 +163,83 @@ test.describe("Login and logout", () => {
...
     );
-    // expect refresh token from token set 2 to match refresh token 2 from request 2 URL
+
+    await fixture.resetClock();
+
+    const authHeader2 = await fixture.getAuthHeader(listenters2.graphql);
```
Comment (mnigh) [inline apps/playwright/tests/login-logout.spec.ts:217]:
```suggestion
    const authHeader2 = await fixture.getAuthHeader(listeners2.graphql);
```

File: apps/playwright/tests/login-logout.spec.ts [hunk 13]
```diff
@@ -171,101 +163,79 @@ test.describe("Login and logout", () => {
...
+
+    const fixture = new AuthTokenFixture(page);
 
     // get auth tokens set 1
-    const tokenSet1 = await getAuthTokens(page);
-    // time travel to when the tokens from token set 1 expire before trying to navigate
-    await clockHelper.jumpTo(jumpPastExpiryDate(tokenSet1?.accessToken ?? ""));
+    const tokenSet1 = await fixture.getTokens();
+
+    // Time travel past the expiry darte of token set 1
```
Comment (mnigh) [inline apps/playwright/tests/login-logout.spec.ts:177]:
```suggestion
    // Time travel past the expiry date of token set 1
```

File: apps/playwright/tests/login-logout.spec.ts [hunk 14]
```diff
@@ -171,101 +163,79 @@ test.describe("Login and logout", () => {
...
+    // Setup request listeners
+    const listeners1 = fixture.createListeners();
+
+    // Reload page to force refresh
+    await fixture.page.reload();
+
+    // Get refresh token used
+    const refreshToken1 = await fixture.getRefreshTokenUsed(listeners1.refresh);
+
+    // Reset clock to avoid unecessary refreshes
```
Comment (mnigh) [inline apps/playwright/tests/login-logout.spec.ts:189]:
```suggestion
    // Reset clock to avoid unnecessary refreshes
```

File: apps/playwright/fixtures/AuthTokenFixture.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (petertgiles) [reply apps/playwright/fixtures/AuthTokenFixture.ts]: Hmm, it certainly seems to work.  You can see the clocking moving forward and backward.  I guess the Date functions inside of the Playwright test itself are not mocked?
image

File: apps/playwright/fixtures/AuthTokenFixture.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (petertgiles) [reply apps/playwright/fixtures/AuthTokenFixture.ts]: Yeah, I like this much better.  :clap:

**Review by petertgiles — APPROVED** [review]
Very cool.  Thanks for cleaning this up.  :pray:
